test/operator.ls

```diff
@@ -88,6 +88,15 @@ a = [1 2 3]
 ok not ("2" in a)
 ok not ("2" in [1 2 3])
 
+# [LiveScript#986](https://github.com/gkz/LiveScript/issues/986)
+# Optimize `in` for one- and zero-element array literals
+eq "x() === 0;"    LiveScript.compile 'x! in [0]'     {+bare,-header}
+eq "x() !== 0;"    LiveScript.compile 'x! not in [0]' {+bare,-header}
+eq "(x(), false);" LiveScript.compile 'x! in [ ]'     {+bare,-header}
+eq "(x(), true);"  LiveScript.compile 'x! not in [ ]' {+bare,-header}
+eq "false;"        LiveScript.compile 'x in [ ]'      {+bare,-header}
+eq "true;"         LiveScript.compile 'x not in [ ]'  {+bare,-header}
```

Comment:
Those two use cases can break if someone defines a property `x` on the global object with a getter having side-effects. I think the optimization should be removed.

Reply:
LiveScript universally (I'm pretty sure) assumes that its users won't do such things. For example, look at the two-element case for this same optimization:
```ls
x in [0 1]
```
compiles to
```js
x === 0 || x === 1;
```
which would also ‘break’ if getting `x` had a side effect by invoking it twice (compare `x! in [0 1]`, which caches a reference to `x!`). It's not just the `in` operator; many kinds of syntax sugar in LiveScript (array/object slices, destructuring, accessignment, the list goes on) make a decision to cache or not cache some expression based on whether it's presumed to possibly have side effects, and simple variable access is never so presumed.

Changing our mind on that is a bigger change than this optimization alone and would merit a bigger discussion. I'm open to that discussion happening, but I don't think it should block this PR, especially since the only non-test code change necessary would be to change the `is-complex` method which this PR uses.

@ymeine, if you're concerned about this behavior, would you start up that discussion in a new issue?

Reply:
Well, given the fact that: 

- first it is quite unlikely to define such a getter on the global object
- then it is quite silly to expect triggering side-effects as a side-effect of an expression itself
- finally I wouldn't write such code which doesn't make much sense except for the reasons you explained as an answer to @vendethiel 

I wouldn't fight too much for this. Also because your arguments make sense.

However I think that for the sake of consistency any LiveScript expression that involves the evaluation of an input expression should evaluate it only once and cache it, no matter how many times it is actually used afterwards to compute the result of the expression. Optimizations make things asymmetrical. Also, the distance between `x!` and `x` is very short, so expecting different behavior in those cases is somewhat disturbing.

I might open an issue later on as you suggested.

Reply:
I think we by and large agree: if you find any instances of LiveScript not compiling to code that evaluates an input expression exactly once, if that expression is anything more complex than an atomic literal or a simple variable, that's a bug. (The caching might not be necessary if the output code naturally uses the expression exactly once, unless the output code would evaluate that expression out of order.) I think the only thing we might disagree on is whether variables should be deprived of their exemption status.

Reply:
Exactly. However your choice of wording made something clear: for variables it's perfectly fine to optimize, for properties it's not. The case I highlighted was for a global property - whose access looks like a variable access - and that can be checked when compiling, maybe it even occurs / would occur in the `is-complex` method.